pulp_ansible/app/tasks/test_tasks.py

```diff
@@ -0,0 +1,94 @@
+                base_path, namespace_name, collection_name
+            )
+            subprocess.run(cmd.split())
+
+            path_to_namespace_dir = os.path.join(base_path, namespace_name)
+            path_to_collection_dir = os.path.join(base_path, namespace_name, collection_name)
+            galaxy_yml_path = os.path.join(path_to_collection_dir, "galaxy.yml")
+            with fileinput.FileInput(galaxy_yml_path, inplace=True) as file:
+                for line in file:
+                    print(line.replace("1.0.0", version), end='')
+
+            cmd = "ansible-galaxy collection build --output-path {} {}".format(
+                base_path, path_to_collection_dir
+            )
+            subprocess.run(cmd.split())
+
+            shutil.rmtree(path_to_namespace_dir)  # Remove the collection build dir
+
+
+def import_collection_from_path(path):
+    artifact = Artifact.init_and_validate(path)
+
+    try:
+        artifact.save()
+    except IntegrityError:
+        artifact = Artifact.objects.get(sha256=artifact.sha256)
+
+    import_collection(artifact.pk)
+
+
+def create_repos_with_collections(repo_count, num_repo_versions, collection_percentage):
+    content_pk = [i for i in CollectionVersion.objects.values_list('pk', flat=True)]
+    last_content_index = math.floor(collection_percentage * len(content_pk))
+
+    for repository_i in range(repo_count):
+        repository, created = AnsibleRepository.objects.get_or_create(name=uuid.uuid4())
+
+        for version_i in range(num_repo_versions):
+            random.shuffle(content_pk)
+            with repository.new_version() as new_version:
+                qs = CollectionVersion.objects.filter(pk__in=content_pk[:last_content_index])
+                new_version.add_content(qs)
+
+
+def promote_content(repos_per_task):
+    # import pydevd_pycharm
+    # pydevd_pycharm.settrace('localhost', port=29437, stdoutToServer=True, stderrToServer=True)
+    random_collection = CollectionVersion.objects.order_by('?').first()
+    repos_to_dispatch = []
+    locks = []
+    for repo in AnsibleRepository.objects.all():
+        repos_to_dispatch.append(repo.pk)
+        locks.append(repo)
+        if len(repos_to_dispatch) == repos_per_task:
+            task_args = (random_collection.pk, repos_to_dispatch)
+            enqueue_with_reservation(promote_content, locks, args=task_args)
+            repos_to_dispatch = []
+            locks = []
+
+    if repos_to_dispatch:
+        task_args = (random_collection.pk, repos_to_dispatch)
+        enqueue_with_reservation(add_content_to_repositories, locks, args=task_args)
+
+
+def add_content_to_repositories(collection_pk, repositories_pks):
+    for repository in AnsibleRepository.objects.filter(pk__in=repositories_pks):
+        with repository.new_version() as new_version:
+            qs = CollectionVersion.objects.filter(pk=collection_pk)
+            new_version.add_content(qs)
```

Comment:
I wonder if Repository.add_content(qs) for a qs of CollectionVersions may be hitting the weakpoints of CollectionVersion (ansible_collectionversion). Ie that each row has a lot of data and a lot of it is in jsonfields and can be a pretty slow to read fully from the db  (ie, `select * from ansible_collectionversion` takes minutes).

Reply:
We merged a performance improvement this morning, and you can see this query analyzed here: https://explain.depesz.com/s/Wi2Q

What is unexpected is that at step 15 is doing Seq Scan, when there is an index available for it. Then I realize though that it's returning enough rows that I think it knows its more optimal to do Seq Scan. When I query for content by pk alone and not as part of `add_content(qs)` it uses IndexScan in that case.